components/script/dom/cookiestore.rs

```diff
@@ -0,0 +1,568 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::borrow::Cow;
+use std::collections::VecDeque;
+use std::rc::Rc;
+
+use base::id::CookieStoreId;
+use cookie::Expiration::DateTime;
+use cookie::{Cookie, SameSite};
+use dom_struct::dom_struct;
+use hyper_serde::Serde;
+use ipc_channel::ipc;
+use ipc_channel::router::ROUTER;
+use itertools::Itertools;
+use js::jsval::NullValue;
+use net_traits::CookieSource::NonHTTP;
+use net_traits::{CookieAsyncResponse, CookieData, CoreResourceMsg, IpcSend};
+use script_bindings::script_runtime::CanGc;
+use servo_url::ServoUrl;
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::CookieStoreBinding::{
+    CookieInit, CookieListItem, CookieSameSite, CookieStoreDeleteOptions, CookieStoreGetOptions,
+    CookieStoreMethods,
+};
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::num::Finite;
+use crate::dom::bindings::refcounted::Trusted;
+use crate::dom::bindings::reflector::{DomGlobal, reflect_dom_object};
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::USVString;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::dom::window::Window;
+use crate::task_source::SendableTaskSource;
+
+/// <https://cookiestore.spec.whatwg.org/>
+/// CookieStore provides an async API for pages and service workers to access and modify cookies.
+/// This requires setting up communication with resource thread's cookie storage that allows for
+/// the page to have multiple cookie storage promises in flight at the same time.
+#[dom_struct]
+pub(crate) struct CookieStore {
+    eventtarget: EventTarget,
+    #[ignore_malloc_size_of = "Rc"]
+    in_flight: DomRefCell<VecDeque<Rc<Promise>>>,
+    // Store an id so that we can send it with requests and the resource thread knows who to respond to
+    #[no_trace]
+    store_id: CookieStoreId,
+}
+
+struct CookieListener {
+    task_source: SendableTaskSource,
```

Comment:
Let's document which task source this is, and add a link to the spec is such an appropriate link exists